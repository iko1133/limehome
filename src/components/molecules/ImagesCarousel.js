import React, { useRef, useState } from "react";
import { Dimensions, Platform } from "react-native";

import styled from "styled-components/native";

const PAGE_WIDTH = Dimensions.get("window").width;

const ImagesScrollContainer = styled.View``;

const ImagesScrollView = styled.ScrollView``;

const Image = styled.Image`
  aspect-ratio: 1.4;
  width: ${PAGE_WIDTH}px;
`;

const PaginationIndicatorContainer = styled.View`
  bottom: 20px;
  flex-direction: row;
  align-self: center;
  position: absolute;
`;

const ImageDot = styled.View`
  background-color: ${({ isSelected }) => (isSelected ? "black" : "white")};
  border: 2px solid white;
  border-radius: 5px;
  height: 8px;
  margin: 0 2px;
  width: 8px;
`;

const ImagesCarousel = ({ images }) => {
  const [shownImageId, setShownImageId] = useState(0);

  const imagesScrollRef = useRef();
  const imagesScrollWidth = useRef(PAGE_WIDTH);
  const imagesScrollPositionX = useRef(0);

  const handleScroll = (event) => {
    const currX = event.nativeEvent.contentOffset.x;

    if (
      Math.round(currX / PAGE_WIDTH) !==
      Math.round(imagesScrollPositionX.current / PAGE_WIDTH)
    )
      setShownImageId(Math.round(currX / PAGE_WIDTH));

    imagesScrollPositionX.current = currX;
  };

  return (
    <ImagesScrollContainer>
      <ImagesScrollView
        ref={imagesScrollRef}
        horizontal
        pagingEnabled
        bounces={false}
        decelerationRate={Platform.OS === "android" ? 0.8 : 0.9}
        showsHorizontalScrollIndicator={false}
        onScroll={(event) => handleScroll(event)}
        scrollEventThrottle={16}
        onContentSizeChange={(w, h) => {
          imagesScrollWidth.current = w;
        }}
      >
        {images?.map((img, index) => (
          <Image key={`product_image_${index}`} source={{ uri: img.url }} />
        ))}
      </ImagesScrollView>

      <PaginationIndicatorContainer>
        {images?.map((img, index) => (
          <ImageDot
            key={`product_image_dot_${index}`}
            isSelected={index === shownImageId}
          />
        ))}
      </PaginationIndicatorContainer>
    </ImagesScrollContainer>
  );
};

export default ImagesCarousel;
