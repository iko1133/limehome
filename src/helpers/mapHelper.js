// Copies given object, except, switches names of the location property's properties for the
// marker objects to function properly
// lgn to longitude
// lat to latitude
export const fixObjectLocationProperty = (object) => ({
  ...object,
  location: {
    ...object.location,
    longitude: object.location?.lng,
    latitude: object.location?.lat,
  },
});
