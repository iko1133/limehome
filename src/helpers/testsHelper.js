import { ThemeProvider } from "styled-components/native";

import { lightTheme } from "styles/theme";

/**
 * Wraps given children into a ThemeProvider with lightTheme set as the main theme
 * @param {object} children - children to be rendered inside ThemeProvider
 */
export const WrapInThemeProvider = (children) => (
  <ThemeProvider theme={lightTheme}>{children}</ThemeProvider>
);
