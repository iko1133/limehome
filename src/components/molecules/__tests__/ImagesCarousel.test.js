import React from "react";
import renderer from "react-test-renderer";

import { WrapInThemeProvider } from "helpers/testsHelper";
import ImagesCarousel from "../ImagesCarousel";

const images = [
  {
    url: "https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/React-icon.svg/1200px-React-icon.svg.png",
  },
  {
    url: "https://www.appstud.com/wp-content/uploads/2018/03/React-Native-Titre.png",
  },
  { url: "https://miro.medium.com/max/800/1*meCFnZ5MK_7Fu1ogYfBvNQ.png" },
];

describe("Testing ImagesCarousel", () => {
  it("Renders images carousel with given one image", () =>
    expect(
      renderer.create(
        WrapInThemeProvider(<ImagesCarousel images={[images[0]]} />)
      )
    ).toMatchSnapshot());

  it("Renders images carousel with all images in it", () =>
    expect(
      renderer.create(WrapInThemeProvider(<ImagesCarousel images={images} />))
    ).toMatchSnapshot());
});
