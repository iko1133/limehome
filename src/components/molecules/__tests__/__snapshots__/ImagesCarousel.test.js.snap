// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ImagesCarousel Renders images carousel with all images in it 1`] = `
<View
  style={
    Array [
      Object {},
    ]
  }
>
  <RCTScrollView
    bounces={false}
    decelerationRate={0.9}
    horizontal={true}
    onContentSizeChange={[Function]}
    onScroll={[Function]}
    pagingEnabled={true}
    scrollEventThrottle={16}
    showsHorizontalScrollIndicator={false}
    style={
      Array [
        Object {},
      ]
    }
  >
    <View>
      <Image
        source={
          Object {
            "uri": "https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/React-icon.svg/1200px-React-icon.svg.png",
          }
        }
        style={
          Array [
            Object {
              "aspectRatio": 1.4,
              "width": 750,
            },
          ]
        }
      />
      <Image
        source={
          Object {
            "uri": "https://www.appstud.com/wp-content/uploads/2018/03/React-Native-Titre.png",
          }
        }
        style={
          Array [
            Object {
              "aspectRatio": 1.4,
              "width": 750,
            },
          ]
        }
      />
      <Image
        source={
          Object {
            "uri": "https://miro.medium.com/max/800/1*meCFnZ5MK_7Fu1ogYfBvNQ.png",
          }
        }
        style={
          Array [
            Object {
              "aspectRatio": 1.4,
              "width": 750,
            },
          ]
        }
      />
    </View>
  </RCTScrollView>
  <View
    style={
      Array [
        Object {
          "alignSelf": "center",
          "bottom": 20,
          "flexDirection": "row",
          "position": "absolute",
        },
      ]
    }
  >
    <View
      isSelected={true}
      style={
        Array [
          Object {
            "backgroundColor": "black",
            "borderColor": "white",
            "borderRadius": 5,
            "borderStyle": "solid",
            "borderWidth": 2,
            "height": 8,
            "marginBottom": 0,
            "marginLeft": 2,
            "marginRight": 2,
            "marginTop": 0,
            "width": 8,
          },
        ]
      }
    />
    <View
      isSelected={false}
      style={
        Array [
          Object {
            "backgroundColor": "white",
            "borderColor": "white",
            "borderRadius": 5,
            "borderStyle": "solid",
            "borderWidth": 2,
            "height": 8,
            "marginBottom": 0,
            "marginLeft": 2,
            "marginRight": 2,
            "marginTop": 0,
            "width": 8,
          },
        ]
      }
    />
    <View
      isSelected={false}
      style={
        Array [
          Object {
            "backgroundColor": "white",
            "borderColor": "white",
            "borderRadius": 5,
            "borderStyle": "solid",
            "borderWidth": 2,
            "height": 8,
            "marginBottom": 0,
            "marginLeft": 2,
            "marginRight": 2,
            "marginTop": 0,
            "width": 8,
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`Testing ImagesCarousel Renders images carousel with given one image 1`] = `
<View
  style={
    Array [
      Object {},
    ]
  }
>
  <RCTScrollView
    bounces={false}
    decelerationRate={0.9}
    horizontal={true}
    onContentSizeChange={[Function]}
    onScroll={[Function]}
    pagingEnabled={true}
    scrollEventThrottle={16}
    showsHorizontalScrollIndicator={false}
    style={
      Array [
        Object {},
      ]
    }
  >
    <View>
      <Image
        source={
          Object {
            "uri": "https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/React-icon.svg/1200px-React-icon.svg.png",
          }
        }
        style={
          Array [
            Object {
              "aspectRatio": 1.4,
              "width": 750,
            },
          ]
        }
      />
    </View>
  </RCTScrollView>
  <View
    style={
      Array [
        Object {
          "alignSelf": "center",
          "bottom": 20,
          "flexDirection": "row",
          "position": "absolute",
        },
      ]
    }
  >
    <View
      isSelected={true}
      style={
        Array [
          Object {
            "backgroundColor": "black",
            "borderColor": "white",
            "borderRadius": 5,
            "borderStyle": "solid",
            "borderWidth": 2,
            "height": 8,
            "marginBottom": 0,
            "marginLeft": 2,
            "marginRight": 2,
            "marginTop": 0,
            "width": 8,
          },
        ]
      }
    />
  </View>
</View>
`;
